Improve robustness for invalid dmidecode characters and add test.

mmulqueen opened the pull request:
Please note, this pull request is identical to https://github.com/saltstack/salt/pull/55450, except that it is based on the master branch instead of the 2018.3 branch.

### What does this PR do?

Ensures that salt does not crash when trying to load DMI data to build the core grains. DMI data is expected to be valid text, however on at least some hardware (Intel NUC marketed around 2014), it contains characters that aren't valid (\xff). This causes salt to crash during start up.

### What issues does this PR fix or reference?

https://github.com/saltstack/salt/issues/48184

### Previous Behavior
With Python 2.7, there is no bug because it was resolved in https://github.com/saltstack/salt/pull/48440

In Python 3, the difference in default behaviour of file opening causes the bug to resurface.

### New Behavior

Files are explicitly opened as bytes in both Python 2.7 and Python 3 versions, this prevents the bug from occurring and makes the previous bug fix work as intended.

### Tests written?
**[NOTICE] Bug fixes or features added to Salt require tests.**
Please review the [test documentation](https://docs.saltstack.com/en/latest/topics/tutorials/writing_tests.html) for details on how to implement tests into Salt's test suite.

Yes. 

Test previously written by @Ch3LL has been adapted and new test written for invalid character. test_core_virtual_unicode and test_core_virtual_invalid

### Commits signed with GPG?

Yes

mmulqueen committed:
Improve robustness for invalid dmidecode characters and add test.

xeacott commented:
Hey @mmulqueen would you be able to look at this PR again and fix the failing checks and resolve the conflicts here?

mmulqueen committed:
Fix coding style (blacken) for DMI fix.

mmulqueen commented:
Hey @xeacott - should be all good now.

xeacott commented:
Perfect, thanks @mmulqueen !

mmulqueen commented:
Not sure what the procedure is here, but I'd like to try to resolve this. When I updated this PR in July, all the tests were passing, including the pre-commit checks. Since then they're started failing, but not for anything connected to this PR. Last week I updated my branch to be up-to-date with master to see whether that would resolve some of the newly failing tests (it did). 

This is a really simple fix, with tests for a problem that crashes the salt-minion on start up - so for the few nodes it affects, it's a pretty big problem.

What can I do to help to get this merged? Preferably without having to refactor it again as it drifts from master.

Ch3LL commented:
apologies on not having this merged yet @mmulqueen 
looks like we will need to get those pre-commit changes updated. If you are not able to do that I am more than willing to update that for you. Just let me know and then we can get this merged in.

Ch3LL committed:
Run pre-commit

mmulqueen commented:
Thanks for merging and fixing the pre-commit. Great to see this go into master.
